tesseract/tesseract_common/include/tesseract_common/clone_cache.h

```diff
@@ -0,0 +1,185 @@
+/**
+ * @file object_cache.h
+ * @brief
+ *
+ * @author Levi Armstrong
+ * @author Matthew Powelson
+ * @date November 15, 2020
+ * @version TODO
+ * @bug No known bugs
+ *
+ * @copyright Copyright (c) 2020, Southwest Research Institute
+ *
+ * @par License
+ * Software License Agreement (Apache License)
+ * @par
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef TESSERACT_COMMON_OBJECT_CACHE_H
+#define TESSERACT_COMMON_OBJECT_CACHE_H
+
+#include <tesseract_common/macros.h>
+TESSERACT_COMMON_IGNORE_WARNINGS_PUSH
+#include <deque>
+#include <memory>
+#include <mutex>
+#include <console_bridge/console.h>
+TESSERACT_COMMON_IGNORE_WARNINGS_POP
+
+#include <tesseract_common/sfinae_utils.h>
+
+namespace tesseract_common
+{
+/** @brief Used to create a cache of objects
+ *
+ * CacheType needs the following methods
+ * CacheType::Ptr clone() const;
+ * int getRevision() const;
+ * bool update(Const CacheType::ConstPtr&);  // optional
+ * */
+template <typename CacheType>
+class CloneCache
+{
+public:
+  CREATE_MEMBER_FUNC_SIGNATURE_CHECK(update, bool, const std::shared_ptr<const CacheType>&)
+
+  CloneCache(std::shared_ptr<CacheType> original, const long& cache_size = 5)
+    : supports_update(has_member_func_signature_update<CacheType>::value)
+    , original_(std::move(original))
+    , cache_size_(static_cast<std::size_t>(cache_size))
+
+  {
+    for (long i = 0; i < cache_size; i++)
```

Comment:
Missing the static assert here using the function generated by `CREATE_MEMBER_FUNC_SIGNATURE_CHECK`.

Reply:
I should be checking it everywhere that it is used. The idea is that it is optional to support updating.

Reply:
So what does the update function do?

Reply:
It updates the object to be at the state of the pattern passed in. So in the case of the environment it would apply any additional commands that have been applied on the environment that's passed in

Reply:
The idea is that we may want caches of things that don't keep a history and can't be updated easily, so for those if they're out of date the only thing you can do is reclone them